ffi/src/log.rs

```diff
@@ -0,0 +1,187 @@
+impl From<Option<&str>> for CStr {
+    fn from(other: Option<&str>) -> Self {
+        if let Some(s) = other {
+            Self::from(s)
+        } else {
+            Self {
+                ptr: std::ptr::null(),
+                len: 0,
+            }
+        }
+    }
+}
+
+/// FFI-safe owned Rust String.
+#[repr(C)]
+pub struct CString {
+    pub ptr: *mut u8,
+    pub cap: usize,
+    pub len: usize,
+}
+
+impl CString {
+    unsafe fn to_str<'a>(&self) -> &'a str {
+        CStr {
+            ptr: self.ptr,
+            len: self.len,
+        }
+        .to_str()
+    }
+}
+
+impl From<String> for CString {
+    fn from(s: String) -> Self {
+        let mut me = ManuallyDrop::new(s);
+        let (ptr, len, cap) = (me.as_mut_ptr(), me.len(), me.capacity());
+        Self { ptr, len, cap }
+    }
+}
+
+impl Drop for CString {
+    fn drop(&mut self) {
+        unsafe {
+            String::from_raw_parts(self.ptr, self.len, self.cap);
+        }
+    }
+}
+
+impl Debug for CString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        unsafe { self.to_str() }.fmt(f)
+    }
+}
+
+impl Debug for CStr {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        unsafe { self.to_str() }.fmt(f)
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct ExternCRecord {
+    pub level: Level,
+    pub message: CString,  // Preformatted message
+    pub module_path: CStr, // None points to null
+    pub file: CStr,        // None points to null
+    pub line: i64,         // None maps to -1, everything else should fit in u32.
+}
+
+impl<'a> From<&Record<'a>> for ExternCRecord {
+    fn from(record: &Record<'a>) -> Self {
+        Self {
+            level: record.level(),
+            message: CString::from(record.args().to_string()),
+            module_path: CStr::from(record.module_path()),
+            file: CStr::from(record.file()),
+            line: record.line().map(|u| u as i64).unwrap_or(-1_i64),
+        }
+    }
+}
+
+static mut LOGGER: Option<ExternCLog> = None;
+static SET_LOGGER: Once = Once::new();
+
+struct ExternCLog {
+    callback: extern "C" fn(&ExternCRecord),
+}
+
+impl ExternCLog {
+    fn new(callback: extern "C" fn(&ExternCRecord)) -> Self {
+        Self { callback }
+    }
+}
+
+impl Log for ExternCLog {
+    fn enabled(&self, _metadata: &Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &Record) {
+        (self.callback)(&ExternCRecord::from(record));
+    }
+
+    fn flush(&self) {}
+}
+
+/// Set the logging callback function that accepts a *const c_char.
+/// The function is idempotent, calling it more then once will have no effect.
+#[no_mangle]
+pub extern "C" fn set_logging_callback(callback: extern "C" fn(&ExternCRecord)) {
+    SET_LOGGER.call_once(|| {
+        unsafe {
+            LOGGER = Some(ExternCLog::new(callback));
+            _ = log::set_logger(LOGGER.as_ref().unwrap());
+        }
+        log::set_max_level(log::LevelFilter::Trace);
+    });
```

Comment:
This means that calling `set_logging_callback` more than once has no effect?

Reply:
Yes. As we talked about offline - I can make it possible to override callback and log level but it would be at the cost of locking a Mutex for each log (because the logger is shared globally). In my opinion, the tradeoff is not worth it because it will not be useful. The most common use case is to set it once forever.

Reply:
Now that I thought about it, do we even need to set a logging level for the Rust code? This will be filtered by the logging level of `zap` inside the callback anyway.

Reply:
We don't have to, but doing it will limit the number of calls back to Go. For example, it's not optimal to call Go callback for every Trace level log if the log level is Info.

Reply:
> at the cost of locking a Mutex for each log (because the logger is shared globally).

That's why I don't like globals 😉. Also the (Rust) logger will never be accessed from more than one go routine anyway (there can only be one instance of `initializer` on the go side). So there is no possibility of a race.